lightning/src/ln/channelmanager.rs

```diff
@@ -1437,121 +1447,141 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-					chacha_stream.read_exact(&mut next_bytes).unwrap();
-					assert_ne!(next_bytes[..], [0; 32][..]);
-					chacha_stream.read_exact(&mut next_bytes).unwrap();
-					assert_ne!(next_bytes[..], [0; 32][..]);
-				}
-
-				// OUR PAYMENT!
-				// final_expiry_too_soon
-				// We have to have some headroom to broadcast on chain if we have the preimage, so make sure we have at least
-				// HTLC_FAIL_BACK_BUFFER blocks to go.
-				// Also, ensure that, in the case of an unknown payment hash, our payment logic has enough time to fail the HTLC backward
-				// before our onchain logic triggers a channel closure (see HTLC_FAIL_BACK_BUFFER rational).
-				if (msg.cltv_expiry as u64) <= self.best_block.read().unwrap().height() as u64 + HTLC_FAIL_BACK_BUFFER as u64 + 1 {
-					return_err!("The final CLTV expiry is too soon to handle", 17, &[0;0]);
-				}
-				// final_incorrect_htlc_amount
-				if next_hop_data.amt_to_forward > msg.amount_msat {
-					return_err!("Upstream node sent less than we were supposed to receive in payment", 19, &byte_utils::be64_to_array(msg.amount_msat));
-				}
-				// final_incorrect_cltv_expiry
-				if next_hop_data.outgoing_cltv_value != msg.cltv_expiry {
-					return_err!("Upstream node set CLTV to the wrong value", 18, &byte_utils::be32_to_array(msg.cltv_expiry));
-				}
-
-				let payment_data = match next_hop_data.format {
-					msgs::OnionHopDataFormat::Legacy { .. } => None,
-					msgs::OnionHopDataFormat::NonFinalNode { .. } => return_err!("Got non final data with an HMAC of 0", 0x4000 | 22, &[0;0]),
-					msgs::OnionHopDataFormat::FinalNode { payment_data } => payment_data,
-				};
+			#[cfg(test)]
+			{
+				// In tests, make sure that the initial onion pcket data is, at least, non-0.
+				// We could do some fancy randomness test here, but, ehh, whatever.
+				// This checks for the issue where you can calculate the path length given the
+				// onion data as all the path entries that the originator sent will be here
+				// as-is (and were originally 0s).
+				// Of course reverse path calculation is still pretty easy given naive routing
+				// algorithms, but this fixes the most-obvious case.
+				let mut next_bytes = [0; 32];
+				chacha_stream.read_exact(&mut next_bytes).unwrap();
+				assert_ne!(next_bytes[..], [0; 32][..]);
+				chacha_stream.read_exact(&mut next_bytes).unwrap();
+				assert_ne!(next_bytes[..], [0; 32][..]);
+			}
 
-				if payment_data.is_none() {
-					return_err!("We require payment_secrets", 0x4000|0x2000|3, &[0;0]);
-				}
+			// OUR PAYMENT!
+			// final_expiry_too_soon
+			// We have to have some headroom to broadcast on chain if we have the preimage, so make sure
+			// we have at least HTLC_FAIL_BACK_BUFFER blocks to go.
+			// Also, ensure that, in the case of an unknown preimage for the received payment hash, our
+			// payment logic has enough time to fail the HTLC backward before our onchain logic triggers a
+			// channel closure (see HTLC_FAIL_BACK_BUFFER rationale).
```

Comment:
>  it's a known defect of our ChannelMonitor logic

If the attack starts with "M2 prevents a channel-close transaction from going onchain because of pinning" I don't think we have a bug, lightning has a bug :).